lib/phoenix_live_view/channel.ex

```diff
@@ -91,19 +123,61 @@ defmodule Phoenix.LiveView.Channel do
     {:noreply, reply(%{state | components: new_components}, msg.ref, :ok, %{})}
   end
 
+  def handle_info(%Message{topic: topic, event: "progress"} = msg, %{topic: topic} = state) do
+    %{socket: socket} = state
+    %{"ref" => ref, "entry_ref" => entry_ref, "progress" => progress} = msg.payload
+    new_socket = Upload.update_progress(socket, ref, entry_ref, progress)
+
+    handle_changed(state, new_socket, msg.ref)
+  end
+
+  def handle_info(%Message{topic: topic, event: "allow_upload"} = msg, %{topic: topic} = state) do
+    %{"ref" => ref} = msg.payload
+    upload_conf = Upload.get_upload_by_ref!(state.socket, ref)
+
+    case Upload.put_entries(state.socket, upload_conf, msg.payload["entries"]) do
+      {:ok, new_socket} ->
+        case Upload.generate_preflight_response(new_socket, upload_conf.name) do
+          {:ok, uploader_reply, new_socket} ->
+            {:noreply, new_state} = handle_changed(state, new_socket, nil)
+            reply(new_state, msg.ref, :ok, uploader_reply)
+            {:noreply, new_state}
+
+          {:error, error_resp, new_socket} ->
+            {:noreply, new_state} = handle_changed(state, new_socket, nil)
+            reply(new_state, msg.ref, :ok, error_resp)
+            {:noreply, new_state}
+        end
+
+      {:error, new_socket, errors} ->
+        errors_reply =
+          Enum.map(errors, fn {ref, msg} -> %{"ref" => ref, "reason" => to_string(msg)} end)
+
+        {:noreply, new_state} = handle_changed(state, new_socket, nil)
+        reply(new_state, msg.ref, :ok, %{error: errors_reply})
+        {:noreply, state}
+    end
```

Comment:
This can be refactored to this:

```suggestion
    %{"ref" => ref, "entries" => entries} = msg.payload
    upload_conf = Upload.get_upload_by_ref!(state.socket, ref)

    {_ok_or_error, reply, new_socket} = 
      case Upload.put_entries(state.socket, upload_conf, entries) do
        {:ok, new_socket} ->
          Upload.generate_preflight_response(new_socket, upload_conf.name)

        {:error, new_socket, errors} ->
          errors_reply =
            Enum.map(errors, fn {ref, msg} -> %{"ref" => ref, "reason" => to_string(msg)} end)
          {:error, new_socket, errors_reply}
      end

    {:noreply, new_state} = handle_changed(state, new_socket, nil)
    reply(new_state, msg.ref, :ok, reply)
    {:noreply, new_state}
```

However I still have some questions:

1. Are we sure the `reply` is always `:ok`?

2. Can we change Upload.put_entries to return an error_response too, like `Upload.generate_preflight_response`? That will help clean up the code a bit further to this:

```elixir
    %{"ref" => ref, "entries" => entries} = msg.payload
    upload_conf = Upload.get_upload_by_ref!(state.socket, ref)

    {_ok_or_error, reply, new_socket} = 
      with {:ok, new_socket} <- Upload.put_entries(state.socket, upload_conf, entries) do
        Upload.generate_preflight_response(new_socket, upload_conf.name)
      end

    {:noreply, new_state} = handle_changed(state, new_socket, nil)
    reply(new_state, msg.ref, :ok, reply)
    {:noreply, new_state}
```

Reply:
1) Yes we always acknowledge w/ `:ok`. 

2) Will dig in more below on the `Upload` feedback then will revisit this Q. I'm sure you're correct :)